src/gmmbin/gmm-init-biphone.cc

```diff
@@ -0,0 +1,245 @@
+        level1_map[pset[k]] = new TableEventMap(0, level2_map);
+
+    } else {
+
+      KALDI_ASSERT(numpdfs_per_phone == 2);
+      int32 base_pdfid = current_pdfid;
+      std::vector<int32> pset = phone_sets[i]; // all these will have a shared event-map child
+      for (size_t k = 0; k < pset.size(); k++) {
+        // create an event map for level2:
+        std::map<EventValueType, EventMap*> level2_map; // key is 0
+        {
+          std::map<EventValueType, EventAnswerType> level3_map; // key is -1
+          level3_map[0] = current_pdfid++;
+          level3_map[1] = current_pdfid++;
+          level2_map[0] = new TableEventMap(kPdfClass, level3_map); // no-left-context case
+        }
+        for (size_t j = 0; j < phone_sets.size(); j++) {
+          std::map<EventValueType, EventAnswerType> level3_map; // key is -1
+          level3_map[0] = current_pdfid++;
+          level3_map[1] = current_pdfid++;
+
+          std::vector<int32> ipset = phone_sets[j]; // all these will have a shared subtree with 2 pdfids
+          for (size_t ik = 0; ik < ipset.size(); ik++) {
+            level2_map[ipset[ik]] = new TableEventMap(kPdfClass, level3_map);
+          }
+        }
+        level1_map[pset[k]] = new TableEventMap(0, level2_map); //lvl2_eventmap;
+        if (k != pset.size() - 1)
+          current_pdfid = base_pdfid;
+      } /////////   k
+
+    }
+  }
+
+  return new TableEventMap(1, level1_map);
+}
+
+
+ContextDependency*
+BiphoneContextDependencyFull(const std::vector<std::vector<int32> > phone_sets,
+                             const std::vector<int32> phone2num_pdf_classes) {
+  std::vector<bool> share_roots(phone_sets.size(), false);  // don't share roots
+  // N is context size, P = position of central phone (must be 0).
+  int32 P = 1, N = 2;
+  EventMap *pdf_map = GetFullBiphoneStubMap(phone_sets,
+                                            phone2num_pdf_classes, share_roots);
+  return new ContextDependency(N, P, pdf_map);
+}
+
+
+} // end namespace kaldi
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using kaldi::int32;
+
+    const char *usage =
+        "Initialize a biphone context-dependency tree with all the\n"
+        "leaves (i.e. a full tree). Intended for end-to-end tree-free models.\n"
+        "Usage:  gmm-init-biphone <topology-in> <dim> <model-out> <tree-out> \n"
+        "e.g.: \n"
+        " gmm-init-biphone topo 39 bi.mdl bi.tree\n";
+
+    bool binary = true;
+    std::string shared_phones_rxfilename;
+    ParseOptions po(usage);
+    po.Register("binary", &binary, "Write output in binary mode");
+    po.Register("shared-phones", &shared_phones_rxfilename,
+                "rxfilename containing, on each line, a list of phones whose pdfs should be shared.");
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 4) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+
+    std::string topo_filename = po.GetArg(1);
+    int dim = atoi(po.GetArg(2).c_str());
```

Comment:
And please use ConvertStringToInteger, e.g. `if (!ConvertStringToInteger(...) || dim <= 0 || dim > 10000) KALDI_ERR...`